ezmlm-dispatch.1
.TH ezmlm-dispatch 1
.SH NAME
ezmlm-dispatch \- Dispatch an email to other ezmlm programs
.SH SYNOPSIS
.B ezmlm-dispatch
[
.B -vV
]
.I dir
[
.I list
]
.SH DESCRIPTION
.B ezmlm-dispatch
is used in a
.I .qmail-default
file to handle delivery to multiple mailing lists in a directory.  It
assumes each list has been created in a directory of the same name as
the list's local name.  It examines the value of
.I $DEFAULT
set by
.B qmail-local
to determine which list the message is destined to, and what file within
the list directory is to control the message delivery.
.P
.RS
.nf
.ta 7c 8c
If the address matches:	Delivery is handled by:
LIST	\fBLIST/editor\fR
LIST-owner	\fBLIST/owner\fR
LIST-digest-owner	\fBLIST/owner\fR
LIST-digest-return-*	\fBLIST/bouncer\fR
LIST-return-*	\fBLIST/bouncer\fR
LIST-confirm-*	\fBLIST/confirmer\fR
LIST-discard-*	\fBLIST/confirmer\fR
LIST-accept-*	\fBLIST/moderator\fR
LIST-reject-*	\fBLIST/moderator\fR
LIST-*	\fBLIST/manager\fR
.fi
.RE
.P
If the optional
.I list
argument is present, it names the list, and
.I $DEFAULT
is only examined for what file is to control the delivery.  This can be
used to have
.B ezmlm-dispatch
control the operations of a single list.
.SH OPTIONS
.TP
.B -v
Display
.B ezmlm-dispatch
version information.
.TP
.B -V
Display
.B ezmlm-dispatch
version information.
.SH RETURN VALUE
.B ezmlm-dispatch
exits with the first non-zero exit code from any of the programs it
executes, or zero otherwise.
.SH SEE ALSO
ezmlm-make(1), ezmlm-manage(1), ezmlm-send(1)
.SH CAVEATS
Unlike
.BR qmail-local ,
.B ezmlm-dispatch
changes directory into the list directory before executing any
commands.  Programs which expect to be executed from the starting
directory will fail.  No ezmlm-idx programs are impacted by this.
.P
For security reasons,
.B ezmlm-dispatch
does not allow delivery to any address with a "/" in it.  This will
break subscribe (and similar) confirmations for senders that use email
addresses with a "/" in them.
.P
In the case of
.B ezmlm-dispatch
handling a single list, it is not possible to use the
.I .qmail-LIST-default
file hand off control to the
.B editor
file for actual list delivery rather than administrative messages.  A
second
.I .qmail-LIST
file will still be necessary in this case.
.SH RESTRICTIONS
.B ezmlm-dispatch
does not support delivery to mbox files or maildirs.
.P
.B ezmlm-dispatch
does not set
.IR $EXT ,
.IR $EXT2 ,
.IR $EXT3 ,
or
.I $EXT4
as
.B qmail-local
does when doing local deliveries.  None of the ezmlm programs use these
environment variables.


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================


ezmlm-dispatch.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "env.h"
#include "sender.h"
#include "strerr.h"
#include "getconfopt.h"
#include "substdio.h"
#include "subfd.h"
#include "messages.h"
#include "error.h"
#include "byte.h"
#include "fmt.h"
#include "str.h"
#include "stralloc.h"
#include "qmail.h"
#include "seek.h"
#include "wrap.h"
#include "slurp.h"
#include "die.h"
#include "auto_version.h"

const char FATAL[] = "ezmlm-dispatch: fatal: ";
const char USAGE[] =
"ezmlm-dispatch: usage: ezmlm-dispatch dir [list]";

static struct option options[] = {
  OPT_END
};

static const char *sender;
static stralloc basedir;
static stralloc path;

static struct qmail qq;
static char strnum[FMT_ULONG];
static int did_program;
static int did_forward;

static const char *makepath(const char *fn)
{
  if (!stralloc_copy(&path,&basedir)) die_nomem();
  if (fn) {
    if (!stralloc_append(&path,"/")) die_nomem();
    if (!stralloc_cats(&path,fn)) die_nomem();
  }
  if (!stralloc_0(&path)) die_nomem();
  return path.s;
}

static int is_dir(const char *fn)
{
  struct stat st;
  return wrap_stat(makepath(fn),&st) == 0
    && S_ISDIR(st.st_mode);
}

static int is_file(const char *fn)
{
  struct stat st;
  return wrap_stat(makepath(fn),&st) == 0
    && S_ISREG(st.st_mode);
}

static void forward(const char *rcpt)
{
  char buf[4096];
  const char *dtline;
  const char *err;
  int r;
  if (qmail_open(&qq) == -1)
    strerr_die2sys(111,FATAL,MSG(ERR_QMAIL_QUEUE));
  if ((dtline = env_get("DTLINE")) != 0)
    qmail_puts(&qq,dtline);
  while ((r = read(0,buf,sizeof buf)) > 0)
    qmail_put(&qq,buf,r);
  if (r == -1)
    strerr_die2sys(111,FATAL,MSG(ERR_READ_STDIN));
  qmail_from(&qq,sender);
  qmail_to(&qq,rcpt);
  if (*(err = qmail_close(&qq)) != '\0')
    strerr_die4x(111,FATAL,MSG(ERR_TMP_QMAIL_QUEUE),": ",err + 1);
  strnum[fmt_ulong(strnum,qmail_qp(&qq))] = 0;
  substdio_puts(subfderr,"qp ");
  substdio_puts(subfderr,strnum);
  substdio_putsflush(subfderr,"\n");
  ++did_forward;
}

static int execute_line(const char *fn,const char *line)
{
  int child;
  int code;
  if (seek_begin(0) == -1)
    strerr_die2sys(111,FATAL,MSG(ERR_SEEK_INPUT));
  while (*line
	 && (*line == '\t'
	     || *line == ' '))
      ++line;
  switch (*line) {
  case '#':
    /* ignore comments */
    return 0;
  case '|':
    if ((child = wrap_fork()) == 0)
      wrap_execsh(line+1);
    code = wrap_waitpid(child);
    ++did_program;
    return code;
  case '/':
  case '.':
    strerr_die5x(100,FATAL,basedir.s,"/",fn,": Delivery to files is not supported.");
    return 100;
  default:
    if (*line == '&')
      ++line;
    forward(line);
    return 0;
  }
}

static int execute_file(const char *fn,stralloc *file)
{
  unsigned int start;
  unsigned int end;
  unsigned int len;
  int code;
  for (start = 0; start < file->len; start = end + 1) {
    len = byte_chr(file->s+start,file->len-start,'\n');
    end = start + len;
    file->s[end] = 0;
    if ((code = execute_line(fn,file->s+start)) != 0)
      return code;
  }
  return 0;
}

static void execute(const char *fn,const char *def)
     /* Load and execute a qmail command file. */
{
  stralloc file = {0,0,0};
  int code;
  if (def != 0)
    env_put2("DEFAULT",def);
  else
    env_unset("DEFAULT");
  fn = makepath(fn);
  if (slurp(fn,&file,256) != 1)
    strerr_die2sys(111,FATAL,MSG1(ERR_READ,fn));
  code = execute_file(fn,&file);
  substdio_puts(subfderr,"did 0+");
  substdio_put(subfderr,strnum,fmt_ulong(strnum,did_forward));
  substdio_puts(subfderr,"+");
  substdio_put(subfderr,strnum,fmt_ulong(strnum,did_program));
  substdio_putsflush(subfderr,"\n");
  _exit(code);
}

static void try_dispatch(const char *def,const char *prefix,unsigned int len,
			 const char *fn)
     /* Test if def has a certain prefix, and if so execute the
      * associated file. */
{
  if (str_diffn(def,prefix,len) == 0
      && is_file(fn))
    execute(fn,def+len);
}

static void dispatch(const char *dir,const char *def)
     /* Hand off control to one of the command files in the list directory. */
{
  if (!is_dir(dir))
    return;
  if (!stralloc_append(&basedir,"/")) die_nomem();
  if (!stralloc_cats(&basedir,dir)) die_nomem();
  /* FIXME: set up $EXT $EXT2 $EXT3 $EXT4 ?  Is it feasable? */
  if (def == 0 || *def == 0)
    execute("editor",0);
  else if (str_diff(def,"owner") == 0)
    execute("owner",0);
  else if (str_diff(def,"digest-owner") == 0)
    execute("owner",0);
  else {
    try_dispatch(def,"digest-return-",14,"digest/bouncer");
    try_dispatch(def,"return-",7,"bouncer");
    try_dispatch(def,"confirm-",8,"confirmer");
    try_dispatch(def,"discard-",8,"confirmer");
    try_dispatch(def,"accept-",7,"moderator");
    try_dispatch(def,"reject-",7,"moderator");
    /* If nothing else matches, call the manager. */
    execute("manager",def);
  }
}

void main(int argc,char **argv)
{
  char *def;
  int dash;
  
  optind = getconfopt(argc,argv,options,0,0);
  if (argv[optind] == 0)
    die_usage();
  if (!stralloc_copys(&basedir,argv[optind++])) die_nomem();

  sender = get_sender();
  if (!sender)
    strerr_die2x(100,FATAL,MSG(ERR_NOSENDER));
  def = env_get("DEFAULT");

  if (argv[optind] != 0) {
    dispatch(argv[optind],def);
    strerr_die3x(100,FATAL,"Not a directory: ",path.s);
  }
  else if (!def || !*def)
    strerr_die2x(100,FATAL,MSG(ERR_NODEFAULT));
  else {
    if (def[str_chr(def,'/')] != 0)
      strerr_die2x(100,FATAL,"Recipient address may not contain '/'");

    dispatch(def,0);

    dash = str_len(def);
    for (;;) {
      while (--dash > 0)
	if (def[dash] == '-')
	  break;
      if (dash <= 0)
	break;
      def[dash] = 0;
      dispatch(def,def+dash+1);
      def[dash] = '-';
    }
    strerr_die3x(100,FATAL,"Could not match recipient name to any list: ",def);
  }
}
